src/compute/src/logging/timely.rs

```diff
@@ -526,3 +325,406 @@ fn create_address_row(id: usize, worker: WorkerIdentifier, address: &[usize]) ->
+        self.output.addresses.give(self.cap, (datum, ts, 1));
+
+        self.state.operators.insert(event.id, event);
+    }
+
+    fn handle_channels(&mut self, event: ChannelsEvent) {
+        let ts = self.ts();
+        let datum = ChannelDatum {
+            id: event.id,
+            source: event.source,
+            target: event.target,
+        };
+        self.output.channels.give(self.cap, (datum, ts, 1));
+
+        let datum = AddressDatum {
+            id: event.id,
+            address: event.scope_addr.clone(),
+        };
+        self.output.addresses.give(self.cap, (datum, ts, 1));
+
+        let dataflow_id = event.scope_addr[0];
+        self.state
+            .dataflow_channels
+            .entry(dataflow_id)
+            .or_default()
+            .push(event);
+    }
+
+    fn handle_shutdown(&mut self, event: ShutdownEvent) {
+        // Dropped operators should result in a negative record for
+        // the `operates` collection, cancelling out the initial
+        // operator announcement.
+        // Remove logging for this operator.
+
+        let Some(operator) = self.state.operators.remove(&event.id) else {
+            error!(operator_id = ?event.id, "missing operator entry at time of shutdown");
+            return;
+        };
+
+        // Retract operator information.
+        let ts = self.ts();
+        let datum = OperateDatum {
+            id: operator.id,
+            name: operator.name,
+        };
+        self.output.operates.give(self.cap, (datum, ts, -1));
+
+        let datum = AddressDatum {
+            id: operator.id,
+            address: operator.addr.clone(),
+        };
+        self.output.addresses.give(self.cap, (datum, ts, -1));
+
+        // Retract schedules information for the operator
+        if let Some(schedules) = self.state.schedules_data.remove(&event.id) {
+            for (bucket, (count, elapsed_ns)) in schedules
+                .into_iter()
+                .enumerate()
+                .filter(|(_, (count, _))| *count != 0)
+            {
+                self.output
+                    .schedules_duration
+                    .give(self.cap, (event.id, ts, -elapsed_ns));
+
+                let datum = ScheduleHistogramDatum {
+                    operator: event.id,
+                    duration_pow: 1 << bucket,
+                };
+                let diff = Diff::cast_from(-count);
+                self.output
+                    .schedules_histogram
+                    .give(self.cap, (datum, ts, diff));
+            }
+        }
+
+        if operator.addr.len() == 1 {
+            let dataflow_id = operator.addr[0];
+            self.handle_dataflow_shutdown(dataflow_id);
+        }
+    }
+
+    fn handle_dataflow_shutdown(&mut self, dataflow_id: usize) {
+        // When a dataflow shuts down, we need to retract all its channels.
+
+        let Some(channels) = self.state.dataflow_channels.remove(&dataflow_id) else {
+            return;
+        };
```

Comment:
Nit: This is subjective, but I would find it more readable here a formulation with `if let Some(channels) = self.state.dataflow_channels.remove(&dataflow_id) {`. Totally optional for your consideration, but it looks like here we are really encoding a conditional, not initialization / error reporting.

Reply:
See my comment above. What I want here is an early return, so this is the equivalent of
```
let channels = self.state.dataflow_channels.remove(&dataflow_id)?;
```

... which we of course can't use because the function doesn't return an `Option`.